(Re-)Add content on generative AI (#697)
swcarpentry/python-novice-gapminder: episodes/04-built-in.md

- Fix syntax errors by reading the source and runtime errors by tracing execution.

## Other ways to get help
There are several other ways that people often get help when they are stuck with their Python code.

* Search the internet: 
  paste the last line of your error message or the word "python" and a short description of what you want to do into your favourite search engine 
  and you will usually find several examples where other people have encountered the same problem and came looking for help.

> * [StackOverflow](https://stackoverflow.com/questions) can be particularly helpful for this: answers to questions are presented as a ranked thread ordered according to how useful other users found them to be.
> * **Take care:** copying and pasting code written by somebody else is risky unless you understand exactly what it is doing!

Reviewer: Are these two points a sublist of "search the internet"?
```suggestion
    * [StackOverflow](https://stackoverflow.com/questions) can be particularly helpful for this: answers to questions are presented as a ranked thread ordered according to how useful other users found them to be.
    * **Take care:** copying and pasting code written by somebody else is risky unless you understand exactly what it is doing!
```

* ask somebody "in the real world". 
  If you have a colleague or friend with more expertise in Python than you have, show them the problem you are having and ask them for help.

> * Sometimes, simply the act of formulating your question can help you to identify what is going wrong.

Reviewer: Isn't the process of talking aloud what helps that process?
```suggestion
* Sometimes, simply the act of formulating your question aloud can help you to identify what is going wrong.
```

Author: Thanks for the suggestion. I will used "articulating" as I hope that will be more inclusive of non-speaking community members.

In addition to asking for help, programmers use generative AI tools to generate code from scratch, extend, improve and reorganise existing code, translate code between programming languages, figure out what terms to use in a search of the internet, and more.
However, there are drawbacks that you should be aware of.

The models used by these tools have been "trained" on very large volumes of data, much of it taken from the internet, and the responses they produce reflect that training data.
Very large amounts of energy was consumed when training most of the models in widespread use and many people are concerned about the environmental cost of this.
Concerns also exist about the way the data for this training was obtained, with questions raised about whether the developers had permission to use it.
Other ethical concerns have also been raised, such as reports that workers were exploited during the training process.

**We recommend that you avoid getting help from generative AI while you learn to code** for several reasons:

> 1. For most problems you will encounter at this stage, help and answers can be easily found by searching the internet.

Reviewer: And if we say "during the workshop" in the paragraph above, then the helpers will be there too.

> 1. For most problems you will encounter at this stage, help and answers can be easily found by searching the internet.

Author: ```suggestion
1. For most problems you will encounter at this stage, help and answers can be found among the first results returned by searching the internet.
```

    * [StackOverflow](https://stackoverflow.com/questions) can be particularly helpful for this: answers to questions are presented as a ranked thread ordered according to how useful other users found them to be.
    * **Take care:** copying and pasting code written by somebody else is risky unless you understand exactly what it is doing!
* ask somebody "in the real world". 
  If you have a colleague or friend with more expertise in Python than you have, show them the problem you are having and ask them for help.
* Sometimes, the act of articulating your question can help you to identify what is going wrong.
  This is known as ["rubber duck debugging"](https://en.wikipedia.org/wiki/Rubber_duck_debugging) among programmers.

### Generative AI

::::::::::::::::::::::::::::: instructor

### Choose how to teach this section
The section on generative AI is intended to be concise but Instructors may choose to devote more time to the topic in a workshop.
Depending on your own level of experience and comfort with talking about and using these tools, you could choose to do any of the following:

> * Explain how large language models work and are trained, and/or the difference between generative AI, other forms of AI that currently exist, and the concept of artificial general intelligence.

Reviewer 2: [nitpick] a discussion of artificial general intelligence seems like it wouldn't be helpful to the learners, in the context of learning how to get help when programming (whereas knowing how LLMs work and are trained would be).  Suggest to delete the last clause:

```suggestion
* Explain how large language models work and are trained, and/or the difference between generative AI and other forms of AI that currently exist.
```

* Demonstrate how you recommend that learners use generative AI.
* Discuss the ethical concerns listed below, as well as others that you are aware of, to help learners make an informed choice about whether or not to use generative AI tools.

This is a fast-moving technology. 
If you are preparing to teach this section and you feel it has become outdated, please open an issue on the lesson repository to let the Maintainers know and/or a pull request to suggest updates and improvements.

> 

Reviewer 3: ```suggestion
If you are comfortable, demonstrating how to work with an LLM, especially through the build in tools, could be a good way to end the workshop, in addition to pointing out other resources. 

```

motivated by @alee's [comment](https://github.com/swcarpentry/python-novice-gapminder/pull/697#discussion_r2002197879)

Author: Can you elaborate on how this differs from the point already included on line 284 of the diff?

> Demonstrate how you recommend that learners use generative AI.

My aim with this guidance to Instructros was to actively encourage different approaches based on the individual's level of expertise and comfort with using the technology

::::::::::::::::::::::::::::::::::::::::

It is increasingly common for people to use _generative AI_ chatbots such as ChatGPT to get help while coding. 

> Once again, you will probably receive some useful guidance by presenting your error message to the chatbot and asking it what went wrong. 

Reviewer 2: [prose/flow] I stumbled at "once again" because it was unclear (to me) what it was referring back to (StackOverflow), until I got 2 sentences further down. Suggest cutting it, the paragraph flows fine without it

```suggestion
You will probably receive some useful guidance by presenting your error message to the chatbot and asking it what went wrong. 
```

However, the way this help is provided by the chatbot is different. 

> Answers on Stackoverflow have (probably) been given by a human as a direct response to the question asked. 

Reviewer 2: [capitalization]

```suggestion
Answers on StackOverflow have (probably) been given by a human as a direct response to the question asked. 
```

But generative AI chatbots, which are based on an advanced statistical model, respond by generating the _most likely_ sequence of text that would follow the prompt they are given.

While responses from generative AI tools can often be helpful, they are not always reliable. 
These tools sometimes generate plausible but incorrect or misleading information, so (just as with an answer found on the internet) it is essential to verify their accuracy.
You need the knowledge and skills to be able to understand these responses, to judge whether or not they are accurate, and to fix any errors in the code it offers you.

> In addition to asking for help, programmers can use generative AI tools to generate code from scratch; extend, improve and reorganise existing code; translate code between programming languages; figure out what terms to use in a search of the internet; and more.

Reviewer 2: this feels like more of a "sidebar" comment, since it's not about getting help *when you're stuck*.  I wonder if it makes more sense to put it at the end of this section?

Author: I intended this to aid flow into the next paragraph. Roughly, something like:

* This is how the thing we are talking about now is similar to the things we talked about just before
* These are some of the ways in which people use this thing beyond what we already talked about
* But these are some ways in which that could be considered problematic

Reviewer 2: I see. In my view, the ethical problems apply regardless of whether you're using an LLM to get help when stuck, or using it to write whole programs from scratch (or refactor, translate code, etc). So I don't see why it needs to be *right here* in that regard.

> However, there are drawbacks that you should be aware of.

Reviewer 2: If you take my suggestion of making the prior paragraph a "sidebar", then this should probably change to something like:

```suggestion
Additionally, there are drawbacks that you should be aware of.
```

> The models used by these tools have been "trained" on very large volumes of data, much of it taken from the internet, and the responses they produce reflect that training data.

Reviewer 2: ```suggestion
The models used by these tools have been "trained" on very large volumes of data, much of it taken from the internet, and the responses they produce reflect that training data, and may recapitulate its inaccuracies or biases.
```

> Very large amounts of energy was consumed when training most of the models in widespread use and many people are concerned about the environmental cost of this.

Reviewer 4: This statement seems a bit out of place and maybe worth cutting, though I do [appreciate the sentiment](https://gerrymcgovern.com/books/world-wide-waste/).

How about something like:

> Training an LLM generally requires vast amounts of energy and computational resources, and issuing a query to LLMs can incur orders of magnitude more energy than a traditional search query [as of this writing](https://www.scientificamerican.com/article/what-do-googles-ai-answers-cost-the-environment/) which may be an ethical consideration to keep in mind as well.

Author: I like this, though I don't think that "incur energy" is quite the right phrasing. Please could you add a suggestion and I will be happy to commit?

Reviewer 5: It's not only the training, it's also the inference. Presumably at some stage all available data will have been used for training so the training costs sink to zero. Inference costs occur every time the model is used.

So maybe just, "Training and using an LLM generally requires...."

Reviewer 3: How about
```suggestion
The environmental costs (energy and water use) of LLMs are a lot higher than other technologies, both during development (known as training) and when an individual user uses one (also called inference). For more information see the [AI Environmental Impact Primer](https://huggingface.co/blog/sasha/ai-environment-primer) developed researchers at HuggingFace, an AI hosting platform. 
```

Reviewer 2: I like @brownsarahm's suggestion. Though it's missing a "by" in "developed [by] researchers at..."

> Concerns also exist about the way the data for this training was obtained, with questions raised about whether the developers had permission to use it.

Reviewer 2: [nitpick] the term "developers" also is used to refer to the people harmed by LLM creation. So I'd suggest referring here to the *companies* to make it less ambiguous.

```suggestion
Concerns also exist about the way the data for this training was obtained, with questions raised about whether the companies developing the LLMs had permission to use it.
```

I'd also consider adding "...permission to use it *in a closed-source revenue-generating product*." But the way I've put it sounds a bit partisan and in-the-weeds. Maybe someone else can think of a better way to capture the crux of the complaint (copyleft license violation) in a succinct and non-technical way.

Other ethical concerns have also been raised, such as reports that workers were exploited during the training process.

**We recommend that you avoid getting help from generative AI during the workshop** for several reasons:

> 1. For most problems you will encounter at this stage, help and answers can be found among the first results returned by searching the internet.

Reviewer 4: I don't really agree with the recommendation that learners not use LLMs during a workshop but this is a pretty soft position. Perhaps the instructor could even use one live and demonstrate the back and forth assessment of the responses provided and techniques for effectively constraining and prompting an LLM. I have found tools like the the latest chatgpt, perplexity, notebooklm, github copilot, v0.dev, etc. to be valuable aids to my work but have also seen first-hand the effects of blind application of LLM responses without critical assessment.

I'm not sure I particularly care for the "cooking" / "recipe" analogy raised elsewhere because in my opinion the main concern at the moment is that LLMs are primarily statistical regurgitation-of-our-past-knowledge black boxes - https://garymarcus.substack.com/p/decoding-and-debunking-hard-forks does a good job of articulating issues around using LLMs for coding.
